packages/react-dom/src/client/DOMPropertyOperations.js

```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
   if (shouldIgnoreAttribute(name, propertyInfo, isCustomComponentTag)) {
     return;
   }
+
+  if (
+    enableCustomElementPropertySupport &&
+    isCustomComponentTag &&
+    name[0] === 'o' &&
+    name[1] === 'n'
+  ) {
+    let eventName = name.replace(/Capture$/, '');
+    const useCapture = name !== eventName;
+    const nameLower = eventName.toLowerCase();
+    if (nameLower in node) {
+      eventName = nameLower;
+    }
+    eventName = eventName.slice(2);
+
+    const listenersObjName = eventName + (useCapture ? 'true' : 'false');
+    const listeners = getCustomElementEventHandlersFromNode(node);
+    const alreadyHadListener = listeners[listenersObjName];
+
+    if (typeof value === 'function' || alreadyHadListener) {
+      listeners[listenersObjName] = value;
+      const proxy = useCapture ? fireEventProxyCapture : fireEventProxy;
+      if (value) {
+        if (!alreadyHadListener) {
+          node.addEventListener(eventName, proxy, useCapture);
+        }
+      } else {
+        node.removeEventListener(eventName, proxy, useCapture);
+      }
+      return;
+    }
+  }
+
   if (shouldRemoveAttribute(name, value, propertyInfo, isCustomComponentTag)) {
     value = null;
   }
+
+  if (
+    enableCustomElementPropertySupport &&
+    isCustomComponentTag &&
```

Comment:
This presumably means this won't work with extended builtins.

If there isn't a good reason for doing so, I would recommend treating custom elements and non-custom elements the same and only special case where they are truly different/special. This will lead to fewer bugs and fewer places where the code diverges into multiple paths.

Reply:
I'm not sure what you mean by extended builtins, but using properties instead of attributes for regular elements in addition to custom elements was never considered before while I've been working on this and sounds like a much bigger change than what I've gotten agreement on so far.

Reply:
You can extend builtin elements like `<button>` which will not have a dasherized name and not be `isCustomComponentTag`. https://developers.google.com/web/fundamentals/web-components/customelements#extendhtml

It's not supported in Safari so not a big deal to not support here, I was just using it as an example of how all of the `isCustomComponentTag` special-casing is going to cause some use-cases to not be covered.

Agree that there is a larger issue to discuss, in another forum. Less special-casing would lead to more uniformity of implementation (easier to maintain) and be easier for users to understand since everything would work the same way.

Reply:
Ah I see, thanks for the clarification! `isCustomComponentTag` appears to include elements with the `is` attribute from the link you shared: https://github.com/facebook/react/blob/b831aec48f2d54913b836d2288b62b7eee9fbca1/packages/react-dom/src/shared/isCustomComponent.js#L12

I am happy to continue this discussion elsewhere as well.